[TG-2721] Use sparse arrays in string_refinementt::get

romainbrenguier opened the pull request:
This makes the interpretation of array coming from the underlying solver more uniform: they are always converted to interval_sparse_arrayt and then accessed using `at` or `concretize`.
Using can reduce a lot the memory consumption in the case of long strings.
We now set a hard limit on the size of strings that are concretized by the solver (for instance to build a trace). 
When this limit is exceeded  CBMC aborts with a message hinting to set `--string-max-input-length` to a reasonable value.
The ultimate goal would be to get rid of `--string-max-length` and only rely on `--string-max-input-length`.

romainbrenguier committed:
Initialize sparse array from array_exprt

romainbrenguier committed:
Initialization of sparse array from array-list

romainbrenguier committed:
Add an interval_sparse_array::of_expr function

This initialize a sparse array from the array expression we know how to
deal with.

romainbrenguier committed:
Tests where model involves long strings

romainbrenguier committed:
Truncate string concatenations in case of overflow

romainbrenguier committed:
Get rid of calls to plus_exprt_with_overflow

This function is adding assumptions on the values of integers which may
lead to contradiction. It is better to deal with overflows at the level
of the specification of the builtin functions instead.

romainbrenguier committed:
Remove plus_exprt_with_overflow_check

romainbrenguier committed:
Use sparse arrays in substitute_array_access

The obtained expression can be exponentially smaller, because sparse
array representation avoids repetitions.

romainbrenguier committed:
Add concretize function for interval_sparse_array

romainbrenguier committed:
Use sparse arrays in get_array

This ensures that arrays from the underlying solver are interpreted in a
consistent manner in the solver (always using interval_sparse_arrayt).

romainbrenguier committed:
Avoid copy in ranged for loops over expressions

romainbrenguier committed:
Remove unused header

romainbrenguier committed:
Use get_array in get_char_array_and_concretize

This makes sure the way we interpret arrays is consistent even in
debugging functions.

romainbrenguier committed:
Stop using concretize_arrays_in_expression

This is now unecessary as get_array takes care of the concretization
using sparse arrays.
This allows use to delete concretize_arrays_in_expression which is now
unused.
Equivalent behavior to this function can be obtained using sparse arrays.

romainbrenguier committed:
Refactor string_refinementt::get

romainbrenguier committed:
Use sparse arrays in string_refinement::get

In the case of index_expressions this can use exponentialy less
memory.

romainbrenguier committed:
Use concretize instead of fill_in_array

This will allow to remove fill_in_array_expr which duplicates what
concretize does.

romainbrenguier committed:
Add a sum_overflows function

Detects when an overflow happens in the sum of two integers.
This will be used in buitin functions for dealing with the overflow case.

romainbrenguier committed:
Add an `at` function for access in sparse arrays

romainbrenguier committed:
Stop using concretize_array_expr in unit tests

We should use sparse arrays as in the solver.

romainbrenguier committed:
Reserve size of array in concretize

romainbrenguier committed:
Clear string_dependencies in calls to dec_solve

If the content is kept between calls to dec_solve, some nodes can be
duplicated in the graph, which leads to more constraints added and
decreased performance.

romainbrenguier committed:
Reduce number of constraints in format

We reduce the number of constraints by avoiding the copy of the last string.

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -152,6 +152,32 @@ exprt interval_sparse_arrayt::to_if_expression(const exprt &index) const
+    if(entries.back().second == expr.operands()[i])
+      entries.back().first = i;
+    else if(entries.back().second.id() == ID_unknown)
+      entries.back() = std::make_pair(i, expr.operands()[i]);
```
There's a more concise way to write this:

```c++
entries.back() = { i, expr.operands()[i] };
```

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -208,6 +218,16 @@ interval_sparse_arrayt::of_expr(const exprt &expr, const exprt &extra_value)
+  for(const auto &pair : entries)
+  {
+    if(pair.first >= index)
+      return pair.second;
```
If it's worth it, you could rewrite `entries` to as `std::map` and then use this function to do this in `O(log n)`:
http://www.cplusplus.com/reference/map/map/upper_bound/

LAJW commented on src/solvers/refinement/string_constraint_generator_format.cpp:
```diff
@@ -392,24 +394,37 @@ exprt string_constraint_generatort::add_axioms_for_format(
+
+  exprt return_code = from_integer(0, get_return_code_type());
+
+  if(intermediary_strings.size() == 0)
```
`if(intermediary_strings.empty())`?

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -228,6 +228,24 @@ exprt interval_sparse_arrayt::at(std::size_t index) const
+  {
+    for(; current_index <= pair.first && current_index < size; ++current_index)
+      array.operands().push_back(pair.second);
+  }
```
I think you could remove `current_index` and add an if statement in this loop:

```c++
if (array.operands().size() < size)
  return array;
```

This should make this function a little bit more obvious.

LAJW commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -935,50 +935,21 @@ static optionalt<exprt> get_array(
+    msg << "consider reducing string-max-input-length so that no string "
+        << "exceeds " << MAX_CONCRETE_STRING_SIZE << " in length and make sure"
+        << " all functions returning strings are available in the classpath";
+    throw msg.str();
```
I'm not a fan of throwing strings. How about `std::runtime_error(string)` or `std::invalid_argument(string)`, if it's not too outlandish of course?

LAJW commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2248,35 +2244,25 @@ exprt string_refinementt::get(const exprt &expr) const
     {
-      auto set = generator.get_created_strings();
-      if(set.find(arr) != set.end())
+      exprt length = super_get(arr.length());
```
`const exprt`?

LAJW commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2234,6 +2234,41 @@ exprt string_refinementt::get(const exprt &expr) const
+    {
+      const auto &if_expr = expr_dynamic_cast<if_exprt>(current.get());
+      const exprt cond = get(if_expr.cond());
+      if(cond.is_true())
```
`if(get(if_expr.cond()).is_true())` - I like oneliners.

romainbrenguier replied:
we also use `cond` with `.is_false` after

LAJW replied:
👌

LAJW reviewed (approved):
I'm so happy to see the concretization and `plus_with_overflow` go!

I've just got a couple suggestions, not even nitpicks, the code looks really good as it is.

romainbrenguier committed:
Refactor interval_sparse_array::concretize

romainbrenguier committed:
Use map instead of vector for sparse array entries

This avoid having to explicitly sort the entries, and make looking up
for an element more efficient: logarithmic instead of linear.

peterschrammel commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -124,7 +124,19 @@ class interval_sparse_arrayt final : public sparse_arrayt
   /// `extra_value`.
   interval_sparse_arrayt(const array_exprt &expr, const exprt &extra_value);
 
+  /// Initialize an sparse array from an array represented by a list of
```
a sparse array

peterschrammel commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -124,7 +124,19 @@ class interval_sparse_arrayt final : public sparse_arrayt
 
+  /// Initialize an sparse array from an array represented by a list of
+  /// index-value pairs, and setting the default to `extra_value`.
+  /// Indexes must be constant expressions, and negative indexes are ignored.
```
When can negative indices occur?

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -176,6 +176,18 @@ interval_sparse_arrayt interval_sparse_arrayt::of_array_list(
+    return interval_sparse_arrayt(*array_expr, extra_value);
+  if(const auto &with_expr = expr_try_dynamic_cast<with_exprt>(expr))
+    return interval_sparse_arrayt(*with_expr);
+  if(expr.id() == "array-list")
```
A proper irep_id should be added for this

romainbrenguier replied:
If that's OK I'd prefer to delay that to anoter PR (prepared here https://github.com/romainbrenguier/cbmc/pull/7) because it will require other reviewers as it touch other parts of CBMC as well.

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -648,20 +648,24 @@ decision_proceduret::resultt string_refinementt::dec_solve()
   output_equations(debug(), equations, ns);
 #endif
 
+  // Dependencies is also used by get, so we have to use it as a class member
```
`dependencies`?

peterschrammel replied:
comment for commit message: `more constraints added and decreased performance`

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -935,50 +935,21 @@ static optionalt<exprt> get_array(
-    return {};
+    std::ostringstream msg;
+    msg << "consider reducing string-max-input-length so that no string "
+        << "exceeds " << MAX_CONCRETE_STRING_SIZE << " in length and make sure"
```
`std::to_string` should be sufficient

romainbrenguier replied:
yes but I don't see the advantage. I think streams are nicer in this kind of situations.

peterschrammel commented on regression/jbmc-strings/long_string/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+Test.class
+--refine-strings --function Test.check --string-max-input-length 2000000
```
How long do these tests take?

romainbrenguier replied:
On my computer 0.7s for the first one and 0.1s for the second.
